ArmPkg/Drivers/SmmuDxe/SmmuDxe.c

```diff
@@ -0,0 +1,774 @@
+  //
+  OutputAddressWidth = SmmuV3DecodeAddressWidth (Idr5.Oas);
+
+  if (OutputAddressWidth < 48) {
+    StreamEntry->S2Ps = SmmuV3EncodeAddressWidth (OutputAddressWidth);
+  } else {
+    StreamEntry->S2Ps = SmmuV3EncodeAddressWidth (48);
+  }
+
+  InputSize           = OutputAddressWidth;
+  StreamEntry->S2T0Sz = 64 - InputSize;
+  if (IortCohac != 0) {
+    StreamEntry->S2Ir0 = ARM64_RGNCACHEATTR_WRITEBACK_WRITEALLOCATE;
+    StreamEntry->S2Or0 = ARM64_RGNCACHEATTR_WRITEBACK_WRITEALLOCATE;
+    StreamEntry->S2Sh0 = ARM64_SHATTR_INNER_SHAREABLE;
+  } else {
+    StreamEntry->S2Ir0 = ARM64_RGNCACHEATTR_NONCACHEABLE;
+    StreamEntry->S2Or0 = ARM64_RGNCACHEATTR_NONCACHEABLE;
+    StreamEntry->S2Sh0 = ARM64_SHATTR_OUTER_SHAREABLE;
+  }
+
+  StreamEntry->S2Rs = 0x2;   // record faults
+
+  if (Idr1.AttrTypesOvr != 0) {
+    StreamEntry->ShCfg = 0x1;
+  }
+
+  if ((Idr1.AttrTypesOvr != 0) && ((CCA == 1) && (CPM == 1) && (DACS == 0))) {
+    StreamEntry->Mtcfg   = 0x1;
+    StreamEntry->MemAttr = 0xF;   // Inner+Outer write-back cached
+    StreamEntry->ShCfg   = 0x3;   // Inner shareable
+  }
+
+  StreamEntry->Valid = 1;
+
+  Status = EFI_SUCCESS;
+  return Status;
+}
+
+/**
+  Allocate a stream table for SMMUv3.
+
+  @param [in]  SmmuInfo       Pointer to the SMMU_INFO structure.
+  @param [in]  SmmuConfig     Pointer to the SMMU configuration.
+  @param [out] Log2Size       Pointer to store the log2 size of the stream table.
+  @param [out] Size           Pointer to store the size of the stream table.
+
+  @retval Pointer to the allocated stream table.
+**/
+SMMUV3_STREAM_TABLE_ENTRY *
+EFIAPI
+SmmuV3AllocateStreamTable (
+  IN SMMU_INFO    *SmmuInfo,
+  IN SMMU_CONFIG  *SmmuConfig,
+  OUT UINT32      *Log2Size,
+  OUT UINT32      *Size
+  )
+{
+  UINT32  MaxStreamId;
+  UINT32  SidMsb;
+  UINT32  Alignment;
+  UINTN   Pages;
+  VOID    *AllocatedAddress;
+
+  MaxStreamId      = SmmuConfig->Config.SmmuNode.SmmuIdMap.OutputBase + SmmuConfig->Config.SmmuNode.SmmuIdMap.NumIds;
+  SidMsb           = HighBitSet32 (MaxStreamId);
+  *Log2Size        = SidMsb + 1;
+  *Size            = SMMUV3_LINEAR_STREAM_TABLE_SIZE_FROM_LOG2 (*Log2Size);
+  *Size            = ROUND_UP (*Size, EFI_PAGE_SIZE);
+  Alignment        = ALIGN_UP_BY (*Size, EFI_PAGE_SIZE);
+  Pages            = EFI_SIZE_TO_PAGES (*Size);
+  AllocatedAddress = AllocateAlignedPages (Pages, Alignment);
+
+  ASSERT (AllocatedAddress != NULL);
+  ZeroMem (AllocatedAddress, *Size);
+  return (SMMUV3_STREAM_TABLE_ENTRY *)AllocatedAddress;
+}
+
+/**
+  Free the allocated stream table for SMMUv3.
+
+  @param [in] StreamTablePtr  Pointer to the stream table entry.
+  @param [in] Size            Size of the stream table.
+**/
+VOID
+EFIAPI
+SmmuV3FreeStreamTable (
+  IN SMMUV3_STREAM_TABLE_ENTRY  *StreamTablePtr,
+  IN UINT32                     Size
+  )
+{
+  UINTN  Pages;
+
+  Pages = EFI_SIZE_TO_PAGES (Size);
+  FreeAlignedPages ((VOID *)StreamTablePtr, Pages);

```

Comment:
Same comment on aligned functions, unless actually required.

Reply:
same comment as above on linear stream tables needed to be aligned to their size.